### ISSUE TITLE
[MSVC] XGBoost failed to build with msvc due to error C2794, error C2938, error C2062, error C3376 and error C2131 with option /std:c++latest on Windows

### ISSUE THREAD

**Issue by Zhaojun-Liu** [dmlc/xgboost#9127]
Hi All,
XGBoost failed to build with msvc with option /std:c++latest on Windows, error message as follows, could you please take a look? Thanks.
```
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\__msvc_iter_core.hpp(140,93): error C2794: 'value_type': is not a member of any direct or indirect base class of 'std::indirectly_readable_traits<xgboost::common::IndexTransformIter<xgboost::linalg::cbegin::<lambda_1>>>' (compiling source file F:\gitP\dmlc\xgboost\src\tree\updater_approx.cc) [F:\gitP\dmlc\xgboost\build_amd64\src\objxgboost.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\xutility(1026,23): error C2938: 'std::iter_value_t' : Failed to specialize alias template (compiling source file F:\gitP\dmlc\xgboost\src\tree\updater_approx.cc) [F:\gitP\dmlc\xgboost\build_amd64\src\objxgboost.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\xutility(5332,8): error C2062: type 'unknown-type' unexpected (compiling source file F:\gitP\dmlc\xgboost\src\tree\updater_approx.cc) [F:\gitP\dmlc\xgboost\build_amd64\src\objxgboost.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\xutility(5332,1): error C3376: 'std::_Equal_memcmp_is_safe_helper': only static data member templates are allowed (compiling source file F:\gitP\dmlc\xgboost\src\tree\updater_approx.cc) [F:\gitP\dmlc\xgboost\build_amd64\src\objxgboost.vcxproj]
 C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\xutility(5336,5): error C2131: expression did not evaluate to a constant (compiling source file F:\gitP\dmlc\xgboost\src\tree\updater_approx.cc) [F:\gitP\dmlc\xgboost\build_amd64\src\objxgboost.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\xutility(5362,19): error C2131: expression did not evaluate to a constant (compiling source file F:\gitP\dmlc\xgboost\src\tree\updater_approx.cc) [F:\gitP\dmlc\xgboost\build_amd64\src\objxgboost.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\__msvc_iter_core.hpp(140,93): error C2794: 'value_type': is not a member of any direct or indirect base class of 'std::indirectly_readable_traits<xgboost::common::IndexTransformIter<xgboost::linalg::cbegin::<lambda_1>>>' (compiling source file F:\gitP\dmlc\xgboost\src\tree\updater_quantile_hist.cc) [F:\gitP\dmlc\xgboost\build_amd64\src\objxgboost.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\xutility(1026,23): error C2938: 'std::iter_value_t' : Failed to specialize alias template (compiling source file F:\gitP\dmlc\xgboost\src\tree\updater_quantile_hist.cc) [F:\gitP\dmlc\xgboost\build_amd64\src\objxgboost.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\xutility(5332,8): error C2062: type 'unknown-type' unexpected (compiling source file F:\gitP\dmlc\xgboost\src\tree\updater_quantile_hist.cc) [F:\gitP\dmlc\xgboost\build_amd64\src\objxgboost.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\xutility(5332,1): error C3376: 'std::_Equal_memcmp_is_safe_helper': only static data member templates are allowed (compiling source file F:\gitP\dmlc\xgboost\src\tree\updater_quantile_hist.cc) [F:\gitP\dmlc\xgboost\build_amd64\src\objxgboost.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\xutility(5336,5): error C2131: expression did not evaluate to a constant (compiling source file F:\gitP\dmlc\xgboost\src\tree\updater_quantile_hist.cc) [F:\gitP\dmlc\xgboost\build_amd64\src\objxgboost.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\xutility(5362,19): error C2131: expression did not evaluate to a constant (compiling source file F:\gitP\dmlc\xgboost\src\tree\updater_quantile_hist.cc) [F:\gitP\dmlc\xgboost\build_amd64\src\objxgboost.vcxproj]
```
**Repro steps:**
1. git clone https://github.com/dmlc/xgboost F:\gitP\dmlc\xgboost
2. git -C "F:\gitP\dmlc\xgboost" reset --hard 36263dd
3. git -C "F:\gitP\dmlc\xgboost" submodule sync
4. git -C "F:\gitP\dmlc\xgboost" submodule foreach git reset --hard
5. git -C "F:\gitP\dmlc\xgboost" submodule foreach git clean -xdf
6. git -C "F:\gitP\dmlc\xgboost" submodule update --init --recursive
7. set _CL_= /std:c++latest
8. mkdir F:\gitP\dmlc\xgboost\build_amd64 and cd F:\gitP\dmlc\xgboost\build_amd64
9. cmake -G "Visual Studio 17 2022" -A x64 -DCMAKE_SYSTEM_VERSION=10.0.22621.0 -DGOOGLE_TEST=ON -DUSE_DMLC_GTEST=ON ..
10. set _CL_=%_CL_% /D_HAS_DEPRECATED_RESULT_OF=1 /D_SILENCE_CXX17_RESULT_OF_DEPRECATION_WARNING /D_HAS_DEPRECATED_SHARED_PTR_UNIQUE=1 /Zc:char8_t-
11. msbuild /m /p:Platform=x64 /p:Configuration=Release xgboost.sln /t:Rebuild

**More:**
I tried the latest commit: a075aa24bad8, the issue still can be reproduced.

**Expected result:**
build successfully.

**Comment by hcho3** [comment]
We don't yet support the flag /c++latest. Our standard build uses C++17 standard.

**Comment by StephanTLavavej** [comment]
I found that I was able to get this to compile by updating `IndexTransformIter`:

* Changing `reference` to be the result of calling `Fn` (i.e. the type currently being used as `value_type`).
* Changing `value_type` to be the result of calling `Fn`, followed by removing reference qualifiers, followed by removing cv-qualifiers (as an iterator's `value_type` should be the "plain" element type)
* Changing the dereference and subscript operators to return `reference` instead of `value_type`

I can't confirm whether this fix is correct and complete, but it certainly appears to fix the compilation errors.

I can submit a PR if you want.
